File: src/control-center/IFetchModule.m
#import <UIKit/UIKit.h>
#import <ControlCenterUIKit/CCUIContentModule-Protocol.h>
#import <ControlCenterUIKit/CCUIContentModuleContentViewController-Protocol.h>
#import <objc/message.h>

#import "../core/IFDiagnostics.h"
#import "../core/IFetchCore.h"

static NSString *IFCCText(NSString *english, NSString *russian) {
    return [IFLanguageManager english:english russian:russian];
}

@interface IFCCMetricView : UIView
@property (nonatomic, strong) UIImageView *metricIconView;
@property (nonatomic, strong) UILabel *valueLabel;
@property (nonatomic, strong) UIProgressView *progressView;
- (instancetype)initWithSymbol:(NSString *)symbol color:(UIColor *)color;
- (void)setValue:(double)value;
@end

@implementation IFCCMetricView

- (instancetype)initWithSymbol:(NSString *)symbol color:(UIColor *)color {
    self = [super init];
    if (self) {
        self.translatesAutoresizingMaskIntoConstraints = NO;
        self.backgroundColor = [UIColor colorWithRed:0.08 green:0.10 blue:0.15 alpha:0.96];
        self.layer.cornerRadius = 11;
        self.layer.borderWidth = 0.5;
        self.layer.borderColor = [UIColor colorWithWhite:1 alpha:0.10].CGColor;
        UIImageSymbolConfiguration *configuration =
            [UIImageSymbolConfiguration configurationWithPointSize:11 weight:UIImageSymbolWeightSemibold];
        UIImage *image = [UIImage systemImageNamed:symbol withConfiguration:configuration];
        self.metricIconView = [[UIImageView alloc] initWithImage:image];
        self.metricIconView.translatesAutoresizingMaskIntoConstraints = NO;
        self.metricIconView.tintColor = color;
        self.metricIconView.contentMode = UIViewContentModeScaleAspectFit;
        self.valueLabel = [[UILabel alloc] init];
        self.valueLabel.translatesAutoresizingMaskIntoConstraints = NO;
        self.valueLabel.font = [UIFont monospacedDigitSystemFontOfSize:14 weight:UIFontWeightBold];
        self.valueLabel.textColor = UIColor.whiteColor;
        self.valueLabel.textAlignment = NSTextAlignmentRight;
        self.progressView = [[UIProgressView alloc] initWithProgressViewStyle:UIProgressViewStyleDefault];
        self.progressView.translatesAutoresizingMaskIntoConstraints = NO;
        self.progressView.progressTintColor = color;
        self.progressView.trackTintColor = [UIColor colorWithWhite:1 alpha:0.14];
        [self addSubview:self.metricIconView];
        [self addSubview:self.valueLabel];
        [self addSubview:self.progressView];
        [NSLayoutConstraint activateConstraints:@[
            [self.metricIconView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor constant:9],
            [self.metricIconView.topAnchor constraintEqualToAnchor:self.topAnchor constant:6],
            [self.metricIconView.widthAnchor constraintEqualToConstant:14],
            [self.metricIconView.heightAnchor constraintEqualToConstant:14],
            [self.valueLabel.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-9],
            [self.valueLabel.centerYAnchor constraintEqualToAnchor:self.metricIconView.centerYAnchor],
            [self.valueLabel.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.metricIconView.trailingAnchor constant:4],
            [self.progressView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor constant:9],
            [self.progressView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-9],
            [self.progressView.topAnchor constraintEqualToAnchor:self.metricIconView.bottomAnchor constant:5],
            [self.progressView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor constant:-7]
        ]];
    }
    return self;
}

- (void)setValue:(double)value {
    double normalized = MIN(100.0, MAX(0.0, value));
    self.valueLabel.text = [NSString stringWithFormat:@"%.0f%%", normalized];
    [self.progressView setProgress:(float)(normalized / 100.0) animated:YES];
}

@end

@interface IFetchModuleViewController : UIViewController <CCUIContentModuleContentViewController>
@property (nonatomic, strong) CAGradientLayer *gradient;
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *subtitleLabel;
@property (nonatomic, strong) UILabel *temperatureLabel;
@property (nonatomic, strong) UILabel *networkLabel;
@property (nonatomic, strong) UIImageView *iconView;
@property (nonatomic, strong) IFCCMetricView *cpuView;
@property (nonatomic, strong) IFCCMetricView *memoryView;
@property (nonatomic, strong) IFLiveMetricsMonitor *monitor;
@property (nonatomic, strong) NSTimer *timer;
@end

@implementation IFetchModuleViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.monitor = [[IFLiveMetricsMonitor alloc] init];
    self.view.clipsToBounds = YES;
    self.view.layer.cornerRadius = 20;
    self.gradient = [CAGradientLayer layer];
    self.gradient.colors = @[
        (id)[UIColor colorWithRed:0.025 green:0.03 blue:0.045 alpha:1].CGColor,
        (id)[UIColor colorWithRed:0.065 green:0.085 blue:0.13 alpha:1].CGColor
    ];
    self.gradient.startPoint = CGPointMake(0, 0);
    self.gradient.endPoint = CGPointMake(1, 1);
    [self.view.layer insertSublayer:self.gradient atIndex:0];

    UIView *iconBackground = [[UIView alloc] init];
    iconBackground.translatesAutoresizingMaskIntoConstraints = NO;
    iconBackground.backgroundColor = [UIColor colorWithRed:0.08 green:0.11 blue:0.17 alpha:1];
    iconBackground.layer.cornerRadius = 12;
    self.iconView = [[UIImageView alloc] initWithImage:[UIImage systemImageNamed:@"waveform.path.ecg.rectangle.fill"]];
    self.iconView.translatesAutoresizingMaskIntoConstraints = NO;
    self.iconView.tintColor = [UIColor colorWithRed:0.28 green:0.62 blue:1 alpha:1];
    [iconBackground addSubview:self.iconView];

    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.titleLabel.text = @"iFetch";
    self.titleLabel.font = [UIFont systemFontOfSize:13 weight:UIFontWeightBold];
    self.titleLabel.textColor = UIColor.whiteColor;

    self.subtitleLabel = [[UILabel alloc] init];
    self.subtitleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.subtitleLabel.font = [UIFont systemFontOfSize:8 weight:UIFontWeightMedium];
    self.subtitleLabel.textColor = [UIColor colorWithWhite:1 alpha:0.64];

    self.temperatureLabel = [[UILabel alloc] init];
    self.temperatureLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.temperatureLabel.font = [UIFont monospacedDigitSystemFontOfSize:11 weight:UIFontWeightSemibold];
    self.temperatureLabel.textColor = UIColor.whiteColor;
    self.temperatureLabel.textAlignment = NSTextAlignmentRight;

    self.cpuView = [[IFCCMetricView alloc] initWithSymbol:@"cpu" color:UIColor.systemOrangeColor];
    self.memoryView = [[IFCCMetricView alloc] initWithSymbol:@"memorychip"
                                                       color:[UIColor colorWithRed:0.70 green:0.48 blue:1 alpha:1]];

    UIView *networkBackground = [[UIView alloc] init];
    networkBackground.translatesAutoresizingMaskIntoConstraints = NO;
    networkBackground.backgroundColor = [UIColor colorWithRed:0.08 green:0.10 blue:0.15 alpha:0.96];
    networkBackground.layer.cornerRadius = 10;
    networkBackground.layer.borderWidth = 0.5;
    networkBackground.layer.borderColor = [UIColor colorWithWhite:1 alpha:0.10].CGColor;
    UIImageView *networkIcon = [[UIImageView alloc] initWithImage:[UIImage systemImageNamed:@"arrow.up.arrow.down"]];
    networkIcon.translatesAutoresizingMaskIntoConstraints = NO;
    networkIcon.tintColor = [UIColor colorWithRed:0.28 green:0.62 blue:1 alpha:1];
    self.networkLabel = [[UILabel alloc] init];
    self.networkLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.networkLabel.font = [UIFont monospacedDigitSystemFontOfSize:9 weight:UIFontWeightSemibold];
    self.networkLabel.textColor = UIColor.whiteColor;
    self.networkLabel.adjustsFontSizeToFitWidth = YES;
    self.networkLabel.minimumScaleFactor = 0.7;
    [networkBackground addSubview:networkIcon];
    [networkBackground addSubview:self.networkLabel];

    [self.view addSubview:iconBackground];
    [self.view addSubview:self.titleLabel];
    [self.view addSubview:self.subtitleLabel];
    [self.view addSubview:self.temperatureLabel];
    [self.view addSubview:self.cpuView];
    [self.view addSubview:self.memoryView];
    [self.view addSubview:networkBackground];

    [NSLayoutConstraint activateConstraints:@[
        [iconBackground.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:10],
        [iconBackground.topAnchor constraintEqualToAnchor:self.view.topAnchor constant:9],
        [iconBackground.widthAnchor constraintEqualToConstant:24],
        [iconBackground.heightAnchor constraintEqualToConstant:24],
        [self.iconView.centerXAnchor constraintEqualToAnchor:iconBackground.centerXAnchor],
        [self.iconView.centerYAnchor constraintEqualToAnchor:iconBackground.centerYAnchor],
        [self.iconView.widthAnchor constraintEqualToConstant:15],
        [self.iconView.heightAnchor constraintEqualToConstant:15],
        [self.titleLabel.leadingAnchor constraintEqualToAnchor:iconBackground.trailingAnchor constant:7],
        [self.titleLabel.topAnchor constraintEqualToAnchor:iconBackground.topAnchor constant:-1],
        [self.subtitleLabel.leadingAnchor constraintEqualToAnchor:self.titleLabel.leadingAnchor],
        [self.subtitleLabel.topAnchor constraintEqualToAnchor:self.titleLabel.bottomAnchor constant:-1],
        [self.temperatureLabel.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-10],
        [self.temperatureLabel.centerYAnchor constraintEqualToAnchor:iconBackground.centerYAnchor],
        [self.temperatureLabel.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.titleLabel.trailingAnchor constant:5],
        [self.cpuView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:9],
        [self.cpuView.topAnchor constraintEqualToAnchor:iconBackground.bottomAnchor constant:7],
        [self.cpuView.heightAnchor constraintEqualToConstant:39],
        [self.memoryView.leadingAnchor constraintEqualToAnchor:self.cpuView.trailingAnchor constant:6],
        [self.memoryView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-9],
        [self.memoryView.topAnchor constraintEqualToAnchor:self.cpuView.topAnchor],
        [self.memoryView.widthAnchor constraintEqualToAnchor:self.cpuView.widthAnchor],
        [self.memoryView.heightAnchor constraintEqualToAnchor:self.cpuView.heightAnchor],
        [networkBackground.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:9],
        [networkBackground.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-9],
        [networkBackground.topAnchor constraintEqualToAnchor:self.cpuView.bottomAnchor constant:6],
        [networkBackground.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor constant:-9],
        [networkIcon.leadingAnchor constraintEqualToAnchor:networkBackground.leadingAnchor constant:8],
        [networkIcon.centerYAnchor constraintEqualToAnchor:networkBackground.centerYAnchor],
        [networkIcon.widthAnchor constraintEqualToConstant:13],
        [networkIcon.heightAnchor constraintEqualToConstant:13],
        [self.networkLabel.leadingAnchor constraintEqualToAnchor:networkIcon.trailingAnchor constant:6],
        [self.networkLabel.trailingAnchor constraintEqualToAnchor:networkBackground.trailingAnchor constant:-7],
        [self.networkLabel.centerYAnchor constraintEqualToAnchor:networkBackground.centerYAnchor]
    ]];
    [self.view addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(openIFetch)]];
    [self refresh:nil];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    self.gradient.frame = self.view.bounds;
}

- (CGFloat)preferredExpandedContentHeight {
    return 190;
}

- (CGFloat)preferredExpandedContentWidth {
    return 320;
}

- (BOOL)providesOwnPlatter {
    return YES;
}

- (void)controlCenterWillPresent {
    [self refresh:nil];
    [self.timer invalidate];
    self.timer = [NSTimer scheduledTimerWithTimeInterval:1 target:self selector:@selector(refresh:) userInfo:nil repeats:YES];
}

- (void)controlCenterDidDismiss {
    [self.timer invalidate];
    self.timer = nil;
}

- (void)refresh:(__unused NSTimer *)timer {
    [self.monitor refresh];
    IFMetricSample *sample = self.monitor.history.lastObject;
    if (sample == nil) {
        return;
    }
    IFBatteryDetails *battery = [IFDiagnostics batteryDetails];
    [self.cpuView setValue:sample.cpuPercent];
    [self.memoryView setValue:sample.memoryPercent];
    self.subtitleLabel.text = IFCCText(@"LIVE SYSTEM", @"СИСТЕМА");
    self.temperatureLabel.text = battery.temperatureCelsius
        ? [NSString stringWithFormat:@"%.1f°C", battery.temperatureCelsius.doubleValue] : @"—";
    self.networkLabel.text = [NSString stringWithFormat:@"↓ %@   ↑ %@",
        [IFetchCore formatRate:sample.downloadBytesPerSecond],
        [IFetchCore formatRate:sample.uploadBytesPerSecond]];
}

- (void)openIFetch {
    @try {
        Class workspaceClass = NSClassFromString(@"LSApplicationWorkspace");
        id workspace = workspaceClass ? [workspaceClass valueForKey:@"defaultWorkspace"] : nil;
        SEL selector = NSSelectorFromString(@"openApplicationWithBundleID:");
        if (workspace != nil && [workspace respondsToSelector:selector]) {
            ((BOOL (*)(id, SEL, id))objc_msgSend)(workspace, selector, @"com.wee1ka.ifetch");
        }
    } @catch (__unused NSException *exception) {
    }
}

@end

@interface IFetchModule : NSObject <CCUIContentModule>
@property (nonatomic, strong) IFetchModuleViewController *controller;
@end

@implementation IFetchModule

- (instancetype)init {
    self = [super init];
    if (self) {
        _controller = [[IFetchModuleViewController alloc] init];
    }
    return self;
}

- (UIViewController<CCUIContentModuleContentViewController> *)contentViewController {
    return self.controller;
}

- (UIViewController *)backgroundViewController {
    return nil;
}

@end

File: src/control-center/IFetchModule.swift
import UIKit

private func IFCCText(_ english: String, _ russian: String) -> String {
    IFLanguageManager.english(english, russian: russian)
}

private final class IFCCMetricView: UIView {
    private let metricIconView: UIImageView
    private let valueLabel = UILabel()
    private let progressView = UIProgressView(progressViewStyle: .default)

    init(symbol: String, color: UIColor) {
        let configuration = UIImage.SymbolConfiguration(pointSize: 11, weight: .semibold)
        metricIconView = UIImageView(image: UIImage(systemName: symbol, withConfiguration: configuration))
        super.init(frame: .zero)

        translatesAutoresizingMaskIntoConstraints = false
        backgroundColor = UIColor(red: 0.08, green: 0.10, blue: 0.15, alpha: 0.96)
        layer.cornerRadius = 11
        layer.borderWidth = 0.5
        layer.borderColor = UIColor(white: 1, alpha: 0.10).cgColor

        metricIconView.translatesAutoresizingMaskIntoConstraints = false
        metricIconView.tintColor = color
        metricIconView.contentMode = .scaleAspectFit

        valueLabel.translatesAutoresizingMaskIntoConstraints = false
        valueLabel.font = .monospacedDigitSystemFont(ofSize: 14, weight: .bold)
        valueLabel.textColor = .white
        valueLabel.textAlignment = .right
        valueLabel.adjustsFontSizeToFitWidth = true
        valueLabel.minimumScaleFactor = 0.7
        valueLabel.numberOfLines = 1
        valueLabel.setContentCompressionResistancePriority(.required, for: .horizontal)

        progressView.translatesAutoresizingMaskIntoConstraints = false
        progressView.progressTintColor = color
        progressView.trackTintColor = UIColor(white: 1, alpha: 0.14)

        addSubview(metricIconView)
        addSubview(valueLabel)
        addSubview(progressView)

        NSLayoutConstraint.activate([
            metricIconView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 9),
            metricIconView.topAnchor.constraint(equalTo: topAnchor, constant: 6),
            metricIconView.widthAnchor.constraint(equalToConstant: 14),
            metricIconView.heightAnchor.constraint(equalToConstant: 14),
            valueLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -9),
            valueLabel.centerYAnchor.constraint(equalTo: metricIconView.centerYAnchor),
            valueLabel.leadingAnchor.constraint(greaterThanOrEqualTo: metricIconView.trailingAnchor, constant: 4),
            progressView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 9),
            progressView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -9),
            progressView.topAnchor.constraint(equalTo: metricIconView.bottomAnchor, constant: 5),
            progressView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -7)
        ])
    }

    required init?(coder: NSCoder) {
        nil
    }

    func setValue(_ value: Double) {
        let normalized = min(100, max(0, value))
        valueLabel.text = String(format: "%.0f%%", normalized)
        progressView.setProgress(Float(normalized / 100), animated: true)
    }
}

@objc(IFetchModuleViewController)
final class IFetchModuleViewController: UIViewController, CCUIContentModuleContentViewController {
    private let gradient = CAGradientLayer()
    private let titleLabel = UILabel()
    private let subtitleLabel = UILabel()
    private let temperatureLabel = UILabel()
    private let networkLabel = UILabel()
    private let iconView = UIImageView(image: UIImage(systemName: "waveform.path.ecg.rectangle.fill"))
    private let cpuView = IFCCMetricView(symbol: "cpu", color: .systemOrange)
    private let memoryView = IFCCMetricView(
        symbol: "memorychip",
        color: UIColor(red: 0.70, green: 0.48, blue: 1, alpha: 1)
    )
    private let monitor = IFLiveMetricsMonitor()
    private var timer: Timer?

    override func viewDidLoad() {
        super.viewDidLoad()

        view.clipsToBounds = true
        view.layer.cornerRadius = 20
        gradient.colors = [
            UIColor(red: 0.025, green: 0.03, blue: 0.045, alpha: 1).cgColor,
            UIColor(red: 0.065, green: 0.085, blue: 0.13, alpha: 1).cgColor
        ]
        gradient.startPoint = CGPoint(x: 0, y: 0)
        gradient.endPoint = CGPoint(x: 1, y: 1)
        view.layer.insertSublayer(gradient, at: 0)

        let iconBackground = UIView()
        iconBackground.translatesAutoresizingMaskIntoConstraints = false
        iconBackground.backgroundColor = UIColor(red: 0.08, green: 0.11, blue: 0.17, alpha: 1)
        iconBackground.layer.cornerRadius = 12

        iconView.translatesAutoresizingMaskIntoConstraints = false
        iconView.tintColor = UIColor(red: 0.28, green: 0.62, blue: 1, alpha: 1)
        iconBackground.addSubview(iconView)

        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.text = "iFetch"
        titleLabel.font = .systemFont(ofSize: 13, weight: .bold)
        titleLabel.textColor = .white

        subtitleLabel.translatesAutoresizingMaskIntoConstraints = false
        subtitleLabel.font = .systemFont(ofSize: 8, weight: .medium)
        subtitleLabel.textColor = UIColor(white: 1, alpha: 0.64)

        temperatureLabel.translatesAutoresizingMaskIntoConstraints = false
        temperatureLabel.font = .monospacedDigitSystemFont(ofSize: 11, weight: .semibold)
        temperatureLabel.textColor = .white
        temperatureLabel.textAlignment = .right

        let networkBackground = UIView()
        networkBackground.translatesAutoresizingMaskIntoConstraints = false
        networkBackground.backgroundColor = UIColor(red: 0.08, green: 0.10, blue: 0.15, alpha: 0.96)
        networkBackground.layer.cornerRadius = 10
        networkBackground.layer.borderWidth = 0.5
        networkBackground.layer.borderColor = UIColor(white: 1, alpha: 0.10).cgColor

        let networkIcon = UIImageView(image: UIImage(systemName: "arrow.up.arrow.down"))
        networkIcon.translatesAutoresizingMaskIntoConstraints = false
        networkIcon.tintColor = UIColor(red: 0.28, green: 0.62, blue: 1, alpha: 1)

        networkLabel.translatesAutoresizingMaskIntoConstraints = false
        networkLabel.font = .monospacedDigitSystemFont(ofSize: 9, weight: .semibold)
        networkLabel.textColor = .white
        networkLabel.adjustsFontSizeToFitWidth = true
        networkLabel.minimumScaleFactor = 0.7

        networkBackground.addSubview(networkIcon)
        networkBackground.addSubview(networkLabel)
        view.addSubview(iconBackground)
        view.addSubview(titleLabel)
        view.addSubview(subtitleLabel)
        view.addSubview(temperatureLabel)
        view.addSubview(cpuView)
        view.addSubview(memoryView)
        view.addSubview(networkBackground)

        NSLayoutConstraint.activate([
            iconBackground.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 10),
            iconBackground.topAnchor.constraint(equalTo: view.topAnchor, constant: 9),
            iconBackground.widthAnchor.constraint(equalToConstant: 24),
            iconBackground.heightAnchor.constraint(equalToConstant: 24),
            iconView.centerXAnchor.constraint(equalTo: iconBackground.centerXAnchor),
            iconView.centerYAnchor.constraint(equalTo: iconBackground.centerYAnchor),
            iconView.widthAnchor.constraint(equalToConstant: 15),
            iconView.heightAnchor.constraint(equalToConstant: 15),
            titleLabel.leadingAnchor.constraint(equalTo: iconBackground.trailingAnchor, constant: 7),
            titleLabel.topAnchor.constraint(equalTo: iconBackground.topAnchor, constant: -1),
            subtitleLabel.leadingAnchor.constraint(equalTo: titleLabel.leadingAnchor),
            subtitleLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: -1),
            temperatureLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -10),
            temperatureLabel.centerYAnchor.constraint(equalTo: iconBackground.centerYAnchor),
            temperatureLabel.leadingAnchor.constraint(greaterThanOrEqualTo: titleLabel.trailingAnchor, constant: 5),
            cpuView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 9),
            cpuView.topAnchor.constraint(equalTo: iconBackground.bottomAnchor, constant: 7),
            cpuView.heightAnchor.constraint(equalToConstant: 39),
            memoryView.leadingAnchor.constraint(equalTo: cpuView.trailingAnchor, constant: 6),
            memoryView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -9),
            memoryView.topAnchor.constraint(equalTo: cpuView.topAnchor),
            memoryView.widthAnchor.constraint(equalTo: cpuView.widthAnchor),
            memoryView.heightAnchor.constraint(equalTo: cpuView.heightAnchor),
            networkBackground.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 9),
            networkBackground.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -9),
            networkBackground.topAnchor.constraint(equalTo: cpuView.bottomAnchor, constant: 6),
            networkBackground.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -9),
            networkIcon.leadingAnchor.constraint(equalTo: networkBackground.leadingAnchor, constant: 8),
            networkIcon.centerYAnchor.constraint(equalTo: networkBackground.centerYAnchor),
            networkIcon.widthAnchor.constraint(equalToConstant: 13),
            networkIcon.heightAnchor.constraint(equalToConstant: 13),
            networkLabel.leadingAnchor.constraint(equalTo: networkIcon.trailingAnchor, constant: 6),
            networkLabel.trailingAnchor.constraint(equalTo: networkBackground.trailingAnchor, constant: -7),
            networkLabel.centerYAnchor.constraint(equalTo: networkBackground.centerYAnchor)
        ])

        view.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(openIFetch)))
        refresh()
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        gradient.frame = view.bounds
    }

    var preferredExpandedContentHeight: CGFloat {
        190
    }

    var preferredExpandedContentWidth: CGFloat {
        360
    }

    var providesOwnPlatter: Bool {
        true
    }

    @objc func controlCenterWillPresent() {
        refresh()
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in
            self?.refresh()
        }
    }

    @objc func controlCenterDidDismiss() {
        timer?.invalidate()
        timer = nil
    }

    private func refresh() {
        monitor.refresh()
        guard let sample = monitor.history.last else {
            return
        }

        let battery = IFDiagnostics.batteryDetails()
        cpuView.setValue(sample.cpuPercent)
        memoryView.setValue(sample.memoryPercent)
        subtitleLabel.text = IFCCText("LIVE SYSTEM", "СИСТЕМА")
        if let temperature = battery.temperatureCelsius {
            temperatureLabel.text = String(format: "%.1f°C", temperature.doubleValue)
        } else {
            temperatureLabel.text = "—"
        }
        networkLabel.text = "↓ \(IFetchCore.formatRate(sample.downloadBytesPerSecond))   ↑ \(IFetchCore.formatRate(sample.uploadBytesPerSecond))"
    }

    @objc private func openIFetch() {
        IFCCOpenApplication("com.wee1ka.ifetch")
    }

    deinit {
        timer?.invalidate()
    }
}

@objc(IFetchModule)
final class IFetchModule: NSObject, CCUIContentModule {
    private let controller = IFetchModuleViewController()

    var contentViewController: (UIViewController & CCUIContentModuleContentViewController) {
        controller
    }

    var backgroundViewController: UIViewController? {
        nil
    }
}
